bin/generate.py

```diff
@@ -1049,24 +1015,96 @@ def generate_visualization():
+                visualize.InputVisualizer
+            )
+            output_visualizer = problem.visualizer(visualize.OutputVisualizer)
+            if output_visualizer is not None:
+                if out_path.is_file() or problem.settings.ans_is_output or problem.interactive:
+                    if visualizer is None or out_path.is_file():
+                        visualizer = output_visualizer
+
+                    if not out_path.is_file() and problem.settings.ans_is_output:
+                        out_path = ans_path
+
+            if visualizer is None:
+                return True
+
+            visualizer_args = testcase.testdata_yaml_args(visualizer, PrintBar())
+
             visualizer_hash = {
-                "visualizer_hash": t.config.visualizer.hash(),
-                "visualizer": t.config.visualizer.cache_command(),
+                "visualizer_hash": visualizer.hash,
+                "visualizer_args": visualizer_args,
             }
 
             if meta_yaml.get("visualizer_hash") == visualizer_hash:
                 return True
 
-            # Generate visualization
-            t.config.visualizer.run(bar, cwd)
+            for ext in config.KNOWN_VISUALIZER_EXTENSIONS:
+                in_path.with_suffix(ext).unlink(True)
 
-            meta_yaml["visualizer_hash"] = visualizer_hash
-            write_yaml(meta_yaml, meta_path, allow_yamllib=True)
+            if isinstance(visualizer, visualize.InputVisualizer):
+                result = visualizer.run(in_path, ans_path, cwd)
+            else:
+                feedbackdir = in_path.with_suffix(".feedbackdir")
+                feedbackdir.mkdir(parents=True, exist_ok=True)
+                teamimage = feedbackdir / "teamimage"
+                judgeimage = feedbackdir / "judgeimage"
+
+                for ext in config.KNOWN_VISUALIZER_EXTENSIONS:
+                    teamimage.with_suffix(ext).unlink(True)
+                    judgeimage.with_suffix(ext).unlink(True)
+
+                result = visualizer.run(
+                    in_path,
+                    ans_path,
+                    out_path if not problem.interactive else None,
+                    feedbackdir,
+                    visualizer_args,
+                )
+                if result.status:
+                    found = None
+                    for ext in config.KNOWN_VISUALIZER_EXTENSIONS:
+                        file = teamimage.with_suffix(ext)
+                        if file.is_file():
+                            found = file
+                    for ext in config.KNOWN_VISUALIZER_EXTENSIONS:
+                        file = judgeimage.with_suffix(ext)
+                        if file.is_file():
+                            found = file
+                    if found is not None:
+                        found.rename(in_path.with_suffix(found.suffix))
+                        bar.log(f"Using {found.name} as testcase visualization")
```

Comment:
Maybe we can specify that this image is from the output visualizer? Maybe it's not needed though, because if the input visualizer does not exist, there's really only one place where this can come from.

```suggestion
                        bar.log(f"Using {found.name} from output_visualizer as test case visualization")
```